openTSNE/affinity.py

```diff
@@ -255,9 +255,26 @@ def check_perplexity(self, perplexity):
 def build_knn_index(
     data, method, k, metric, metric_params=None, n_jobs=1, random_state=None
 ):
+    if not sp.issparse(data) and metric in [
```

Comment:
This does not check anywhere if annoy is actually installed. I'd start with annoy being an optional dependency. This is easy as well, just add some flag at the top of the file like

```python
try:
    import annoy
    HAS_ANNOY = True
except ImportError:
    HAS_ANNOY = False
```

Reply:
Are you sure about that? Why not make it a compulsory dependency? It's on pip/conda, it's very stable, so there shouldn't be any problems. If you make it optional then most end users will never see the difference because they won't have it installed.

Reply:
I thought about gradually introducing it, in case there are some issues, and until we perform proper benchmarks. But either way, I just want to get this merged into master, so I can play with this on my end. In that case, please add `annoy` to the `setup.py` dependencies. I'll add it to the conda recipe before release.

Reply:
OK. I added it to setup.py and removed from azure-pipelines. You can change it if you prefer.

Reply:
I think it's called `annoy` in pip but `python-annoy` in conda, so I wasn't completely sure what to write in the `setup.py` though. I wrote `annoy`.

Reply:
That's right. Pypi dependencies, which are installed through pip have to go into either the setup.py file or sometimes the requirements.txt file. Conda packages have to be included in the conda recipe, which has its own repository (https://github.com/conda-forge/opentsne-feedstock), so `python-annoy` will go in there.